samples/fpga/gemm_systolic_vectorized.py

```diff
@@ -0,0 +1,430 @@
+    state.add_memlet_path(compute_tasklet,
+                          exit_m,
+                          exit_k,
+                          C_buffer_out,
+                          memlet=dace.Memlet("C_buffer[m]"),
+                          src_conn="c_out")
+    state.add_memlet_path(C_buffer_out, exit_n0, memlet=dace.Memlet())
+
+
+    write_c_tasklet = state.add_tasklet(
+        "write_c", {"buffer_in", "forward_in"}, {"c_out"}, """\
+if n1 <= p:
+    c_out = forward_in if p > 0 and n1 > 0 else buffer_in""")
+    state.add_memlet_path(C_buffer_out,
+                          entry_c,
+                          write_c_tasklet,
+                          memlet=dace.Memlet("C_buffer[m]", dynamic=True),
+                          dst_conn="buffer_in")
+    state.add_memlet_path(C_pipe_in,
+                          entry_n0,
+                          entry_c,
+                          write_c_tasklet,
+                          memlet=dace.Memlet("C_pipe[p-1]", dynamic=True),
+                          dst_conn="forward_in")
+    state.add_memlet_path(write_c_tasklet,
+                          exit_c,
+                          exit_n0,
+                          C_pipe_out,
+                          memlet=dace.Memlet("C_pipe[p]", dynamic=True),
+                          src_conn="c_out")
+
+    # Unroll processing elements
+    compute_entry, compute_exit = state.add_map(
+        "unroll_compute", {"p": "0:P"},
+        schedule=dace.ScheduleType.FPGA_Device,
+        unroll=True)
+
+    # Bring data nodes into scope
+    state.add_memlet_path(compute_entry, A_pipe_in, memlet=dace.memlet.Memlet())
+    state.add_memlet_path(compute_entry, B_pipe_in, memlet=dace.memlet.Memlet())
+    state.add_memlet_path(compute_entry, C_pipe_in, memlet=dace.memlet.Memlet())
+    state.add_memlet_path(A_pipe_out, compute_exit, memlet=dace.memlet.Memlet())
+    state.add_memlet_path(B_pipe_out, compute_exit, memlet=dace.memlet.Memlet())
+    state.add_memlet_path(C_pipe_out, compute_exit, memlet=dace.memlet.Memlet())
+
+
+def make_fpga_state(sdfg, vec_width=1):
+    vec_type = dace.vector(dace.float32, vec_width)
+
+    state = sdfg.add_state("gemm")
+
+    sdfg.add_stream("A_pipe",
+                    dace.float32,
+                    transient=True,
+                    shape=(P + 1, ),
+                    storage=dace.dtypes.StorageType.FPGA_Local,
+                    buffer_size="P")
+    sdfg.add_stream("B_pipe",
+                    vec_type,
+                    transient=True,
+                    shape=(P + 1, ),
+                    storage=dace.dtypes.StorageType.FPGA_Local)
+    sdfg.add_stream("C_pipe",
+                    vec_type,
+                    transient=True,
+                    shape=(P + 1, ),
+                    storage=dace.dtypes.StorageType.FPGA_Local)
+
+    make_read_A(state)
+    make_read_B(state, sdfg, vec_width)
+    make_compute(sdfg, state, vec_width)
+    make_write_C(state, sdfg, vec_width)
```

Comment:
Are you sure you implemented `gemm`? Shouldn't there be a `read_C`?

Reply:
it's in the writer

Reply:
Interesting idea :O Could you update the name?